third_party/move/move-compiler-v2/src/pipeline/instruction_reordering.rs

```diff
@@ -0,0 +1,771 @@
+    /// in which case, the entries are dropped.
+    /// However, `mapping` cannot be missing offsets for non-`Prepare` instructions
+    /// (specifically, offsets for use instructions).
+    pub fn filter_remap(&mut self, mapping: BTreeMap<CodeOffset, CodeOffset>) {
+        self.0 = self
+            .0
+            .iter()
+            .filter_map(|(prepare_offset, use_info)| {
+                mapping.get(prepare_offset).map(|remapped_prepare_offset| {
+                    (*remapped_prepare_offset, UseInfo {
+                        offset: *mapping.get(&use_info.offset).expect("existing offset"),
+                        copy_only: use_info.copy_only,
+                    })
+                })
+            })
+            .collect();
+    }
+
+    /// Extend with the `other` annotation.
+    /// The `base_offset` is added to the instruction offsets in `other`.
+    pub fn extend_with(&mut self, other: Self, base_offset: CodeOffset) {
+        for (prepare_offset, use_info) in other.0.into_iter() {
+            self.0.insert(base_offset + prepare_offset, UseInfo {
+                offset: base_offset + use_info.offset,
+                ..use_info
+            });
+        }
+    }
+}
+
+/// Use information corresponding to a `Prepare` instruction.
+/// It refers to the original instruction for which the `Prepare` instruction
+/// is being inserted.
+#[derive(Clone)]
+pub struct UseInfo {
+    // Offset of the original instruction for which the `Prepare` instruction is inserted.
+    pub offset: CodeOffset,
+    // Instructs the file format generator that when true, the corresponding `Prepare`
+    // can only result in a copy (not a move). If a copy is not possible, then
+    // the `Prepare` instruction should be a no-op.
+    // When false, the `Prepare` instruction can result in a move or a copy.
+    pub copy_only: bool,
+}
+
+/// Hold various information to safely perform reordering of instructions within a block.
+struct BlockReordering {
+    instructions: Vec<Bytecode>,
+    dd_graph: DataDependenceGraph,
+    prepare_use: PrepareUseAnnotation,
+    constraints: BTreeMap<CodeOffset, BTreeSet<CodeOffset>>,
+    original_block_len: usize,
+}
+
+impl BlockReordering {
+    /// Given `instructions` of a basic block, insert required `Prepare` instructions
+    /// and safely reorder the instructions when possible, to be friendlier for
+    /// generating stack machine bytecode.
+    pub fn prepare_and_reorder(instructions: Vec<Bytecode>) -> ReorderedInstructions {
+        let original_block_len = instructions.len();
+        let mut this = Self {
+            instructions,
+            dd_graph: DataDependenceGraph(BTreeMap::new()),
+            prepare_use: PrepareUseAnnotation::default(),
+            constraints: BTreeMap::new(),
+            original_block_len,
+        };
+        this.fill_data_dependence_graph();
+        this.fill_prepare_instructions();
+        this.fill_constraints();
```

Comment:
It seems to me that your `fill_constraints` tries to ignore the inserted `Prepare` instructions.  If that is the case, we can probably simplify that code if we (1) run it before `fill_prepare_instructions`, and then we can (2) allow it to run on the complete set of instructions, not consulting `original_block_len` in each pass.

You can probably then simplify a bit by having `fill_prepare_instructions` return the list of `Prepare` instructions, rather than adding them to `instructions`, and pass it as a parameter to `reorder_constructions`.